tex/generic/pgf/systemlayer/pgfsys-dvips.def

```diff
@@ -159,28 +164,491 @@
   }%
-  \ht#1=0pt
-  \dp#1=0pt
-  \hskip\pgf@picminx\hbox to 0pt{\box#1\hss}%
-  \special{ps:
-    grestore restore
+  \wd#1=0pt\ht#1=0pt\dp#1=0pt%
+  \hskip\pgf@picminx\box#1%
+  \pgfsys@outerinvoke{%
+    /pgffoa pgf@sys@foa def /pgfsoa pgf@sys@soa def
+    grestore
     mark /EP pdfmark
-    mark {pgfxform} << /Group << /S /Transparency
+    mark {pgfxform@\the\pgf@objectcount} << /Group << /S /Transparency
       /I \ifpgfsys@transparency@group@isolated true \else false \fi
-      /K \ifpgfsys@transparency@group@knockout true \else false \fi  >> >> /PUT pdfmark
+      /K \ifpgfsys@transparency@group@knockout true \else false \fi  >> >> /PUT pdfmark%
+  }%
+  \setbox#1=\hbox{\pgfsys@outerinvoke{mark {pgfxform@\the\pgf@objectcount} /SP pdfmark}}%
+}
+
+
+%
+% Image inclusion, adding SMask reference and image interpolation if mandated
+%
+
+\def\pgfsys@defineimage{%
+  \ifx\pgf@imagewidth\pgfutil@empty\else\edef\pgf@imagewidth{width=\pgf@imagewidth,}\fi%
+  \ifx\pgf@imageheight\pgfutil@empty\else\edef\pgf@imageheight{height=\pgf@imageheight,}\fi%
+  \begingroup%
+  \def\pgf@imageaddedkeys{}\def\@tempa{true}%
+  \ifx\pgf@imageinterpolate\@tempa\def\pgf@imageaddedkeys{/Interpolate true}\fi%
+  \ifx\pgf@imagemask\pgfutil@empty\else\edef\pgf@imageaddedkeys{\pgf@imageaddedkeys /SMask {\pgf@imagemask}}\fi%
+  \edef\@tempa{%
+    \ifx\pgf@imageaddedkeys\pgfutil@empty\else%
+      \noexpand\pgfsys@outerinvoke{%
+        save /pgfparent [(img) cvx cvn rand] cvx def %random name for current image
+        mark /_objdef /pgfparent load /NI pdfmark
+        % sometimes, image metadata contain embedded pdfmarks that prevent us from
+        % writing to the image dictionary ourselves, so we disable them temporarily
+        save /pdfmark /cleartomark load def
+      }%
+    \fi%
+    \noexpand\includegraphics[\pgf@imageheight\pgf@imagewidth]{\pgf@filename}%
+    \ifx\pgf@imageaddedkeys\pgfutil@empty\else%
+      \noexpand\pgfsys@outerinvoke{%
+        % empty image dummy, in case the provided file is not a valid raster image
+        0 0 1 [1 0 0 1 0 0] {} image restore
+        mark /pgfparent load <<\pgf@imageaddedkeys>> /PUT pdfmark restore
+      }%
+    \fi%
+  }%
+  \expandafter\endgroup%
+  \expandafter\def\expandafter\pgf@image\expandafter{\@tempa}%
+}
+
+
+%
+% Image masks
+%
+
+\def\pgfsys@definemask#1{%
+  \global\advance\pgf@objectcount\@ne%
+  \ifx\pgf@maskmatte\pgfutil@empty%
+  \else%
+    \edef\pgf@maskmatte{/Matte [\pgf@maskmatte]}%
+  \fi%
+  \pgfsys@outerinvoke{%
+    mark /_objdef {pgfmask\the\pgf@objectcount} /NI pdfmark
+    1 dict begin /pdfmark /cleartomark load def
+  }%
+  \special{psfile=\pgf@filename\space hsize=0 vsize=0}%
+  \pgfsys@outerinvoke{%
+    0 0 1 [1 0 0 1 0 0] {} image
+    end mark {pgfmask\the\pgf@objectcount} <</Decode [1 0] \pgf@maskmatte>> /PUT pdfmark
   }%
-  \setbox#1=\hbox{\special{ps: mark {pgfxform} /SP pdfmark mark /NamespacePop pdfmark}}%
+  \edef\pgf@mask{pgfmask\the\pgf@objectcount}%
 }
```

Comment:
This macro doesn't use its argument.  Is this really correct?

Reply:
That's correct.

In the `pdftex` driver too, there is *no* use made of the box number passed as argument to the macro. Instead, a PDF XForm object is generated directly from the file content (`pdftex`: command `\pdfximage`, `dvips`: `NI` pdfmark) to be referenced later by its object number. Btw, none of the remaining drivers defining this macro (`dvipdfmx`, `luatex`) make use of `#1`.

Reply:
I see.  Once this is merged you could refactor all the drivers to (and possibly the documentation) to remove the argument entirely.